proto/view.proto

```diff
@@ -0,0 +1,54 @@
+// Copyright 2014 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tamir Duberstein (tamird@gmail.com)
+// Author: Vivek Menezes (vivek.menezes@gmail.com)
+//
+// All the protobuf messages that are used to represent a database View.
+// A View can be a join of many tables, but currently can only reference
+// a single table.
+
+syntax = "proto2";
+package cockroach.proto;
+option go_package = "proto";
+
+import "gogoproto/gogo.proto";
+
+option (gogoproto.sizer_all) = true;
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+
+// Column is the column metadata provided in a ResultSet.
+message Column {
+  // A name should encode a namespace, table, and column name.
+  repeated bytes name = 1;
+  // Column type, size, etc to be put here.
+}
+
+// Result is one Result in a ResultSet for a database View.
+message Result {
+  repeated bytes columns = 1;
+}
+
+// ResultSet is a slice of a database View. A View of the database can be
+// very large and is to be accommodated in a stream of ResultSet
+message ResultSet {
+  // The columns that form the view.
+  repeated Column columns = 1;
+  // Some of the results in a View. All the results are for the same columns.
+  repeated Result results = 3;
+}
```

Comment:
I'd like us to iron out the protobuf API to be as close to the final
outcome so we don't have to keep changing all the key value clients
significantly
every time we change this API. Besides it's nice to see the compromises we
need to make on behalf of a key value only client when they use a view
approach.

We do want ResultSet to be part of the response sent back by Cockroach
right?


On Wed, May 20, 2015 at 11:23 AM Peter Mattis <notifications@github.com>
wrote:

> In proto/view.proto
> <https://github.com/cockroachdb/cockroach/pull/1074#discussion_r30713789>:
>
> > +  // Column type, size, etc to be put here.
> > +}
> > +
> > +// Result is one Result in a ResultSet for a database View.
> > +message Result {
> > +  repeated bytes columns = 1;
> > +}
> > +
> > +// ResultSet is a slice of a database View. A View of the database can be
> > +// very large and is to be accommodated in a stream of ResultSet
> > +message ResultSet {
> > +  // The columns that form the view.
> > +  repeated Column columns = 1;
> > +  // Some of the results in a View. All the results are for the same columns.
> > +  repeated Result results = 3;
> > +}
>
> This (Column, Result and ResultSet) feels a bit premature. It isn't
> needed for a "key-value" table. And we first need to iron out the
> NamespaceDescriptor and TableDescriptor. Perhaps I'm missing some context
> on why this comes first.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1074/files#r30713789>.
>